### PR TITLE
C# Proto2 feature : Groups

### PR DESCRIPTION
**Pull request by ObsidianMinor**
Another day, another proto2 PR for C#.

This PR adds support for groups (including unknown field support) in the C# library and code generator.

### CODE REVIEW

**Comment by ObsidianMinor** [comment]
I don't believe the failed ruby build is due to what I wrote, @anandolee.

File: csharp/src/Google.Protobuf/CodedInputStream.cs [hunk 1]
```diff
@@ -383,6 +384,10 @@ public uint ReadTag()
                 // If we actually read a tag with a field of 0, that's not a valid tag.
                 throw InvalidProtocolBufferException.InvalidTag();
             }
+            if (ReachedLimit)
```
[the remark this answers is not in the document]
Comment (ObsidianMinor) [reply csharp/src/Google.Protobuf/CodedInputStream.cs]: Actually, that is supposed to also have a ReachedTagLimit check but I forgot to add it in when moving code.

File: src/google/protobuf/compiler/csharp/csharp_group_field.cc [hunk 2]
```diff
@@ -0,0 +1,188 @@
+// Protocol Buffers - Google's data interchange format
```
Comment (anandolee) [inline src/google/protobuf/compiler/csharp/csharp_group_field.cc]: Both c++ and java do not have generators for group:
https://github.com/protocolbuffers/protobuf/tree/ba8692fbade4ba329cc4531e286ab5a8e0821d97/src/google/protobuf/compiler/java
https://github.com/protocolbuffers/protobuf/tree/ba8692fbade4ba329cc4531e286ab5a8e0821d97/src/google/protobuf/compiler/cpp

The group can share code gen with messages:
https://github.com/protocolbuffers/protobuf/blob/ba8692fbade4ba329cc4531e286ab5a8e0821d97/src/google/protobuf/compiler/java/java_helpers.cc#L373

**Comment by ObsidianMinor** [comment]
Fixed. I also include a fix for an issue that would prevent group end tags from being written in repeated fields.

**Comment by ObsidianMinor** [comment]
Fixed the other build files containing remnants of the separate group generators

**Comment by anandolee** [comment]
No worries, it is normal to fixing the build file several times :)

**Pull request merged**